Fix Unity8 tests again (again?)

UniversalSuperBox opened the pull request:
Merry Christmas, everyone! For your present, I fixed the build. Ignore the fact that I broke the build. *whistles nonchalantly*

Just some more race conditions and interesting behavior. In particular, the TopLevelWindowModel is capable of issuing ~~1 million~~ (edit: now 2 billion) unique window IDs but instead sometimes issues the same ID for multiple windows. Fun! With some atomicity added, this case should happen much less often now.

Then we continue on, fix a couple of longer-standing problems in the test suite and do some refactoring to help reduce some race conditions leading to tests that fail *sometimes*. Finish off by quieting down a ton of warnings from fake indicators and we finish with a slightly more sane test battery.

Please either **merge with a merge commit** or **rebase and merge**, the history is meaningful.

UniversalSuperBox committed:
Reduce warnings caused by fake indicators

UniversalSuperBox committed:
Make touchFlick flick instead of snap

Prior to this change, NaN was stored in 'distance' and every touchFlick
would snap the touch point from the start to the end instantly.

After fixing this, touchFlick took much longer, so I also increased the
default speed.

The only test that this broke was test_rightEdgePushWithOpenIndicators,
so I replaced the touchFlick with showIndicators(). While I was in the
file I left it a little nicer than I came by placing all utility
functions above the test functions.

UniversalSuperBox committed:
Make it less likely to issue duplicate Window IDs

Some tests will issue the null window and the first window at almost the
same time, causing their window IDs to be the same prior to this change.
That was fun to debug.

I'm not going to pretend that this is absolutely foolproof, it's
possible we still end up with duplicate IDs, but the risk should be
quite small.

I had trouble testing this with the rest of the PhoneStage tests where
it is now customary to put the TLWM tests, there is too much troublesome
behavior when changing maxWindowId with an in-use model. Instead, the
new tst_QMLTopLevelWindowModel gives a case similar to PhoneStage but
with less touch input and more raw window actions.

UniversalSuperBox committed:
Fix name of QMLTopLevelWindowModel tests

UniversalSuperBox committed:
Improve showing launcher automatically

If the Panel is open and the Greeter is swiped away, the Panel was
automatically dismissed by the Launcher appearing. Avoid this
(in most cases) by making sure the Panel isn't visible at all before
calling the Launcher.

A warning was generated in the tests when we tried to check
!greeter.active since greeter was null.

UniversalSuperBox committed:
Refactor and fix intermittent tests

Some tests in tst_Shell would fail due to latent conditions, mostly due
to trying to open the Panel while the Launcher was animated. The
Launcher's animations block Panel actions.

To fix this, we can load the desktop stage with at least one app. This
prevents the Launcher from being called and helps the situation immensely.

While I was at it, I refactored the loadDesktopShellWithApps function to
allow loading fewer than the maximum number of apps, and to avoid using
unity8-dash since it has special behavior in the tests that we don't
want to test for any more. Then I used the refactored function
everywhere that we manually loaded the desktop shell before.

And for the cherry on top, I fixed a couple 'could not read property of
null' errors from the test controls.

UniversalSuperBox committed:
Remove panel visibility check from launcher

Trying to avoid race conditions, I wanted to prevent the launcher from
dismissing the Panel when the Shell was unlocked. Turns out the Panel is
*always* set to visible when unlocking, so the Launcher never shows.

Ref 00c396672f4bd1a2abec0f4bbddfd5d6e682df0e

UniversalSuperBox commented:
*On the 25th day of Christmas, Jenkins CI said to me,*
`This commit cannot be built`

UniversalSuperBox committed:
Basically skip animations if Showable not shown

There are instances where a Showable may be told to hide even if it is
not currently shown. This happens quite a lot in use, since just saying
'hide()' is far better than 'if (shown) hide()'. However, this causes
the Showable's hideAnimation to play, even if it's already hidden.

This can cause an interesting problem when items' usability depends on
whether their animations are playing, such as the PanelMenus disabling
their DragHandles[1].

To fix this, immediately complete the hideAnimation (and showAnimation)
when they are not needed. Doing this instead of not playing the
animations at all corrects situations where shown was not set but the
Showable is being shown somehow.

This caused the panel tests to fail on timing issues (attempting to
start a touch event while the Panel was still disabled), so I had to
refactor one and add an ugly wait().

This also caused ShellWithPin's test_bfbOnLockedDevice to fail since the
shell tried to open the Drawer then immediately call launcher.hide().
Previously the timing worked out somehow and hide() was called first. To
fix this, remove the onSessionStarted() handler and move it to the
Greeter's onActiveChanged handler instead.

[1] https://github.com/ubports/unity8/pull/205

UniversalSuperBox committed:
Add copyright headers I missed

UniversalSuperBox committed:
Keep refactoring tst_Shell, reduce race conditions

Refactor loadDesktopShellWithApps into loadShell to have a single place
to fix problems with shell or app loading in this test case.

Use the new refactored function everywhere.

I had to shuffle the order of setting up the greeter to cope with the
change, and add some more waitForRendering magic to make sure we don't
stomp on the Shell while it loads.

UniversalSuperBox commented:
Finally, all passing. Ready for review.

dobey commented on plugins/WindowManager/TopLevelWindowModel.cpp:
```diff
@@ -38,7 +39,8 @@ Q_LOGGING_CATEGORY(TOPLEVELWINDOWMODEL, "toplevelwindowmodel", QtInfoMsg)
 TopLevelWindowModel::TopLevelWindowModel()
-    : m_nullWindow(createWindow(nullptr)),
+    : m_nullWindow(createNullWindow()),
+      m_maxWindowId(1000000),
```
I presume this is changed to a non-const int to be able to initialize or set to a very small value for tests. However, the old comment below is I think wrong and should be removed, and we should use `std::numeric_limits<int>::max()` here, as Qt seems to use a 32-bit int for the QML int.

UniversalSuperBox replied:
It is to allow for messing with it in tests, yes, and now I see the QML documentation says "The possible int values range from around -2000000000 to around 2000000000", the range of a 32-bit signed integer. I'll make that change.

UniversalSuperBox replied:
Got it, and tests passed locally. Now we wait for `-j1` tests to do their work...

UniversalSuperBox committed:
Use the max value of an int for max window ID

The QML documentation says "The possible int values range from around
-2000000000 to around 2000000000", the range of a 32-bit signed integer.
This makes our old comment lie, we can use the max int value for
the platform.

mariogrip reviewed (approved):
LGTM :+1:
